Fixing extended trace failure for Adam and AdaMax and generalising `alpha` parameter to accept callable object (scheduler)

kishore-nori opened the pull request:
This is a small PR for Adam and AdaMax methods (thanks a lot for adding them to `Optim.jl`) to 

- fix issue https://github.com/JuliaNLSolvers/Optim.jl/issues/1096 (when using `extended_trace=true`) by adding `alpha` to the respective `State` structs, so that `common_trace!` works out of the box.
- generalise `alpha` (and effected `update_state!`) for both `Adam` and `AdaMax` such that they can accept a callable object - a function or in particular a scheduler. This helps in using the functionality from [ParameterSchedulers.jl](https://github.com/FluxML/ParameterSchedulers.jl) seamlessly, without adding it as a dependency.

pkofod commented:
I am fine with these changes and thank you for fixing the original error. I do have to ask you to update the docstrings of the types to reflect the new feature and also add tests for the bug that was fixed as well as the new feature.

Thanks!

kishore-nori commented:
Thank you for checking the PR and for the comments. I have added:

- details in the docstrings of Adam and AdaMax for scheduled alpha constructor
- added tests for Adam and AdaMax,  covering the `extended_trace=true` option that was failing (https://github.com/JuliaNLSolvers/Optim.jl/issues/1096) 
- added tests for scheduled Adam and AdaMax convergence,  and tests covering the `extended_trace=true` option to check if the alpha values are correct.

pkofod commented:
Thank you

kishore-nori commented:
I am glad, would be happy to know if you recommend any corrections or changes to make it merge ready. 

Is it possible to run the CI workflow tests to check if the tests work fine? (locally they ran successfully)

pkofod commented:
> I am glad, would be happy to know if you recommend any corrections or changes to make it merge ready.
> 
> Is it possible to run the CI workflow tests to check if the tests work fine? (locally they ran successfully)

Running now :)

pkofod commented:
Seem to fail, not sure why it failed in a newton test.

kishore-nori commented:
Thanks a lot for running the CI. It unclear to me as well as to why Newton tests failed, and that too on MacOS. I ll investigate and get back.

pkofod commented:
Not sure what happened to be honest.. This time it worked

kishore-nori commented:
That's great! Thank you for re-running. Where can I check the earlier CI failure?

kishore-nori commented:
Ok I found the earlier CI runs to trace the failure: CI says `Paraboloid Diagonal` failed, which I think is the below test:

https://github.com/JuliaNLSolvers/OptimTestProblems.jl/blob/c1fba66c90b44934d13cd11b2502573f1c11fab8/src/optim_tests/multivariate/quad_transforms.jl#L107-L142

I have a feeling `guardseed(0)` failed to accomplish what it is supposed to and the random matrix used in the test could have been a "bad" one. But I am not sure if this problem could come from architecture or MacOS, or even "just" a failure of `guardseed`. However, shouldn't be related to the changes in this PR. 

some related issues:

https://github.com/JuliaLang/julia/issues/42752 and https://github.com/JuliaLang/julia/issues/51225

pkofod commented:
Thank you for looking into it. We'll have to see if it reappears.

kishore-nori commented:
Is the PR good go?

kishore-nori commented:
sorry to bother, but is the PR good go? or do I need to improve it?

pkofod commented:
Sorry, this had slipped from my attention, will look at it again and merge if good :) Thanks!

kishore-nori commented:
No worries, thank you!

pkofod commented:
Looks good. Thanks. I think alpha should maybe use linesearch keywords, but since alpha is already there it's okay with me.

kishore-nori commented:
The tests seem to have failed, but unrelated to this PR.

>  I think alpha should maybe use linesearch keywords, but since alpha is already there it's okay with me.

Yes true, but would probably be a breaking change.
